### PR TITLE
feat: Add keep pairs option for keys deletion

### LINKED ISSUE #1476: jwk: Add keep pairs option for keys delete
**Is your feature request related to a problem? Please describe.**

When rotating JWKs, I delete old keys at first and it takes time and effort like this.

```bash
#!/usr/bin/env sh

hydra keys get hydra.openid.id-token | jq -r '.Payload.keys[].kid' \
| awk 'NR>2 {print}' | sed -e 's/^"//' -e 's/"$//' \
| xargs -I{} hydra keys delete hydra.openid.id-token/{}

hydra keys get hydra.jwt.access-token | jq -r '.Payload.keys[].kid' \
| awk 'NR>2 {print}' | sed -e 's/^"//' -e 's/"$//' \
| xargs -I{} hydra keys delete hydra.jwt.access-token/{}

hydra keys get hydra.https-tls | jq -r '.Payload.keys[].kid' \
| awk 'NR>2 {print}' | sed -e 's/^"//' -e 's/"$//' \
| xargs -I{} hydra keys delete hydra.https-tls/{}
```

This complicated script means delete keys except latest key pairs each JWK Set.
I execute this before rotating system secret because hydra cannot serve if keys associated deleted system secret exists.

**Describe the solution you'd like**

Add `keep-pairs` option to delete a JSON Web Key Set endpoint and command.

For example: 

* API: `/keys/{set}?keep-pairs=1`
* Command: `hydra keys delete {set} --keep-pairs 1`

This is same as following

```bash
hydra keys get {set} | jq -r '.Payload.keys[].kid' \
| awk 'NR>2 {print}' | sed -e 's/^"//' -e 's/"$//' \
| xargs -I{} hydra keys delete {set}/{}
```

**Additional context**

My system rotation step is here.
This proposal helps first step.

1. Delete old JWKs (hydra.openid.id-token/hydra.jwt.access-token/hydra.https-tls)
1. Delete a old system secret
1. Add a new system secret
1. Create JWKs (hydra.openid.id-token/hydra.jwt.access-token/hydra.https-tls)
1. Restart hydra server

### PR DESCRIPTION
**Pull request by sawadashota**
## Related issue

This PR will closes #1476

## Proposed changes

- [x] Add tests for JWK API
- [x] Add keep pairs option for keys deletion API
- [x] Generate SDK 
- [x] Add keep pairs option for keys deletion CLI

## Checklist

- [x] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md)
- [x] I have read the [security policy](../security/policy)
- [x] I confirm that this pull request does not address a security vulnerability. If this pull request addresses a security
vulnerability, I confirm that I got green light (please contact [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push the changes.
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation within the code base (if appropriate)
- [ ] I have documented my changes in the [developer guide](https://github.com/ory/docs) (if appropriate)

### CODE REVIEW

File: jwk/handler.go [hunk 1]
```diff
@@ -327,8 +328,22 @@ func (h *Handler) UpdateKey(w http.ResponseWriter, r *http.Request, ps httproute
 //       500: genericError
 func (h *Handler) DeleteKeySet(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
 	var setName = ps.ByName("set")
+	var olderThan = r.URL.Query().Get("older-than")
 
-	if err := h.r.KeyManager().DeleteKeySet(r.Context(), setName); err != nil {
+	if olderThan != "" {
+		date, err := time.Parse("2006-01-02", olderThan)
```
Comment (aeneasr) [inline jwk/handler.go]: Would a unix timestamp make more sense here? Because here we don't know about timezones etc.

File: jwk/handler.go [hunk 2]
```diff
@@ -327,8 +328,22 @@ func (h *Handler) UpdateKey(w http.ResponseWriter, r *http.Request, ps httproute
 //       500: genericError
 func (h *Handler) DeleteKeySet(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
 	var setName = ps.ByName("set")
+	var olderThan = r.URL.Query().Get("older-than")
```
Comment (aeneasr) [inline jwk/handler.go]: This should be documented in swagger :)

Comment (aeneasr) [reply jwk/handler.go]: Also maybe rename this to `before`?

File: jwk/handler_test.go [hunk 3]
```diff
@@ -70,3 +71,66 @@ func TestHandlerWellKnown(t *testing.T) {
...
+
+		var getJWKSet jose.JSONWebKeySet
+		err = json.NewDecoder(getRes.Body).Decode(&getJWKSet)
+		require.NoError(t, err, "problem in decoding response")
+
+		require.Len(t, getJWKSet.Keys, 2)
+	})
+
+	t.Run("DeleteJSONWebKeySet", func(t *testing.T) {
+		deleteJWKSetPath := "/keys/test-key"
```
Comment (aeneasr) [inline jwk/handler_test.go]: The older-than/before feature should be tested here!

File: jwk/manager_sql.go [hunk 4]
```diff
@@ -211,6 +211,13 @@ func (m *SQLManager) DeleteKey(ctx context.Context, set, kid string) error {
 	return nil
 }
 
+func (m *SQLManager) DeleteOldKeys(ctx context.Context, set string, date time.Time) error {
```
Comment (aeneasr) [inline jwk/manager_sql.go]: The manager_test should have a test for this method

File: jwk/handler.go (hunk 1 above)
Comment (sawadashota) [reply jwk/handler.go]: Sounds better! I fixed it!

File: jwk/handler.go (hunk 2 above)
Comment (sawadashota) [reply jwk/handler.go]: Yes, I added the param to `jwk/docs.go`

File: jwk/handler_test.go (hunk 3 above)
Comment (sawadashota) [reply jwk/handler_test.go]: Sure! I will add a test case :)

File: jwk/manager_sql.go (hunk 4 above)
Comment (sawadashota) [reply jwk/manager_sql.go]: This is also I will do it!

**Comment by aeneasr** [comment]
Let me know when this is good for another review :)

**Comment by sawadashota** [comment]
@aeneasr I have done! Could you review again?

**Comment by sawadashota** [comment]
I couldn't generate SDK before but I get to do it now!
So I added a `--older-than` option  to key deletion.

@aeneasr I'm so sorry to be late. Can you review again?

**Comment by aeneasr** [comment]
Closing due to inactivity - would probably need another overhaul to address this as we've changed quite a lot internally in the meanwhile.

**Pull request closed without merging**